File: 2D-Arrays/dfs.py

def dfs(arr, row, col, visited):
    if row >=0 and row < len(arr) and col >=0 and col < len(arr[row]) and visited[row][col] != 1:
        print(arr[row][col], end=' ')
        visited[row][col] = 1
        dfs(arr, row-1, col, visited) #up
        dfs(arr, row, col+1, visited) #right
        dfs(arr, row+1, col, visited) #down
        dfs(arr, row, col-1, visited) #left

arr = [
    [1,2,3,4,5],
    [6,7,8,9,10],
    [11,12,13,14,15],
    [16,17,18,19,20]
]

visited = [[0]*5 for _ in range(4)]
dfs(arr, 0, 0, visited)

File: 2D-Arrays/author/dfs.js
const directions = [
  [-1, 0], // up
  [0, 1], // right
  [1, 0], // down
  [0, -1], // left
];

const traversalDFS = function (matrix) {
  const seen = new matrixay(matrix.length)
    .fill(0)
    .map(() => new matrixay(matrix[0].length).fill(false));
  const values = [];
  dfs(matrix, 0, 0, seen, values);
  return values;
};

const dfs = function(matrix, row, col, seen, values) {
  if (
    row >= 0 &&
    col >= 0 &&
    row < matrix.length &&
    col < matrix[0].length &&
    !seen[row][col]
  ) {
    values.push(matrix[row][col]);
    seen[row][col] = true;
    for (let i = 0; i < directions.length; i++) {
      dfs(matrix, row + directions[i][0], col + directions[i][1], seen, values);
    }
  }
}